[Yeap Yi Sheng James] iP

jamesyeap opened the pull request:
# Spark
> "Busy is the new Happy" - some workaholic

Are you always busy? 

**AWESOME!** Spark helps you remember just how busy you are by letting you store your todos, deadlines and events all in one place!

Now you'll have solid, indisputable  _documentary proof_ of just how much of a workaholic you are! 😗

Spark is
- text-based
- easy to learn
- **SUPER FAST** to use

## Getting Started
1. download it from [here](https://github.com/jamesyeap/ip/releases/tag/Level-9)
2. place the `.jar` file inside a new folder
3. open up your favourite terminal, and run 
```bash
java -jar spark.jar
```
4. add the endless list of things you have on your plate
5. use it as the ultimate excuse to escape social-interactions by showing it to people!

And it is **FREE**!

## Features
- [x] Managing Todos
- [x] Managing Deadlines
- [x] Managing Events
- [x] Fully usable on the Command-Line Interface (CLI)
- [ ] A Graphical User-Interface (GUI) **_(COMING SOON)_**

Commit:
Add Gradle support

jamesyeap committed:
Rename Duke to Spark.

jamesyeap committed:
Add function to echo user commands and to exit when user's command is "bye".

jamesyeap committed:
Add functions to add task and list all tasks. User's commands will also no longer be echoed.

jamesyeap committed:
Change Task toString() method to reflect its completion status and added setter-functions to mark Task as Done or Not-Done.

jamesyeap committed:
Add taskId attribute to Task

jamesyeap committed:
Add feature to mark and unmark tasks using taskId

jamesyeap committed:
Update README.md

jamesyeap committed:
Add 3 new sub-classes of Task: Todo, Deadline and Event

jamesyeap committed:
Add feature to add 3 different types of Tasks: ToDo, Deadline and Event

jamesyeap committed:
Add feature to reject invalid commands with an error-message.

jamesyeap committed:
Minor formatting change.

jamesyeap committed:
Add feature to show number of tasks in the list after adding a task.

jamesyeap committed:
Add 3 new Exceptions specific to Spark.

jamesyeap committed:
Add function to reject adding a Task with empty field(s).

jamesyeap committed:
Create a base Exception class SparkException.

jamesyeap committed:
Create an Exception class that represents all Exceptions arising from errors in formatting in user-inputs; ie missing Title when creating a new Todo.

jamesyeap committed:
Create Exceptions that are thrown when user adds an invalid Task.

jamesyeap committed:
Create Exception that is thrown when user enters an invalid command.

jamesyeap committed:
Move TaskList, Task and its sub-classes into a package.

jamesyeap committed:
Add error-messages for invalid user inputs.

jamesyeap committed:
Change implementation of code to check if a required field is empty.

jamesyeap committed:
Add function to delete a Task by a given TaskId.

jamesyeap committed:
Remove TaskId attribute from Task.

jamesyeap committed:
Add Exception that is thrown when user enters a non-integer for taskId when marking, unmarking or deleting a Task.

jamesyeap committed:
Add an additional test-case.

jamesyeap committed:
Add Exceptions that are thrown when user marks a Task that is already marked or unmarks a Task that is already unmarked.

jamesyeap committed:
Add Command enums to enable Spark to recognise command keywords that are not completely-lowercase but otherwise valid; e.g. tOdO will now be recognised as a valid keyword to add a new todo task

jamesyeap committed:
Add function for Tasks added to be saved on user's hard-disk.

jamesyeap committed:
Add function for deleted Tasks to be removed from save-file in hard-disk.

jamesyeap committed:
Change storage format of date and time in Event

jamesyeap committed:
Change storage format of date and time in Deadline

jamesyeap committed:
Update input testcases

jamesyeap committed:
Extract user-interface related functions into a class

jamesyeap committed:
Extract parsing of user commands into a class

jamesyeap committed:
Add function to find Tasks with titles that contain specified keywords

jamesyeap committed:
Add JavaDocs for public methods in class TaskList

jamesyeap committed:
Indent long lines that exceed character-limit (>120 characters)

jamesyeap committed:
Merge branch 'branch-A-CodingStandard'

jamesyeap committed:
Merge branch 'branch-A-JavaDoc'

# Conflicts:
#	src/main/java/spark/tasks/TaskList.java

jamesyeap committed:
Update README.md

jamesyeap committed:
Fix welcome message

jamesyeap committed:
Add two unit-test cases for Spark

jamesyeap committed:
Ignore .jar files in Git VCS

jamesyeap committed:
Setup build settings for JAR file creation

jamesyeap committed:
Remove an unnecessary method: Spark#exit()

jamesyeap committed:
Implement function to check validity of params

jamesyeap committed:
Merge remote-tracking branch 'origin/add-gradle-support'

jamesyeap committed:
Add JavaFX dependencies to Gradle

jamesyeap committed:
Add checkstyle config file

jamesyeap committed:
Add non-functional UI for Spark

jamesyeap committed:
Add some styles to the UI for Spark

jamesyeap committed:
Add placeholder response-message in UI

jamesyeap committed:
Change UI layout to suit Spark

jamesyeap committed:
Add exception to catch invalid date input

jamesyeap committed:
Change exception message for modifying non-existent Task

jamesyeap committed:
Merge branch 'master' into branch-level10-gui

jamesyeap committed:
Remove unnecessary lines in Parser

jamesyeap committed:
Add additional JavaDocs

jamesyeap committed:
Add additional JavaDocs

jamesyeap committed:
Fix order of import statements in TaskList

jamesyeap committed:
Change naming of AddToDoCommand to AddTodoCommand

jamesyeap committed:
Add minor changes to Javadocs comments

jamesyeap committed:
Change layout of Spark's UI back to original layout

jamesyeap committed:
Fix bug where input-bot exceeded window width

jamesyeap committed:
Add a semi-working GUI for Spark

jamesyeap committed:
Change avatars for user and spark

jamesyeap committed:
Change background color for dialog boxes

jamesyeap committed:
Change font family and size for all text

Ryan-L98 commented on src/main/java/spark/Spark.java:
```diff
@@ -0,0 +1,92 @@
+import spark.exceptions.fileexceptions.TaskDecodingException;
+
+public class Spark {
+    private static final String defaultFilePathString = "spark_save_file.txt";
```
Constant names must be all uppercase using underscore to separate words! 😆

Ryan-L98 commented on src/main/java/spark/Spark.java:
```diff
@@ -0,0 +1,92 @@
+            this.storage = new Storage(filePathString);
+            this.taskList = new TaskList(storage.readTasksFile());
+        } catch (FileException | TaskDecodingException e) {
+            ui.printException(e);
```
You may consider including the same exception message(line 31) here to be consistent 😄

Ryan-L98 commented on src/main/java/spark/Spark.java:
```diff
@@ -0,0 +1,92 @@
+        try {
+            Command command = Parser.parseInput(userInput);
+            String response = command.execute(taskList, ui, storage);
+            boolean isExit = command.isExit();
```
The boolean variable isExit is not used in this method 🤔

Ryan-L98 commented on src/main/java/spark/Ui.java:
```diff
@@ -0,0 +1,71 @@
+            "What can I do for you?"
+            );
+
+    private final Scanner userInput;
```
Constant names must be all uppercase using underscore to separate words. 😄

zawedcvg replied:
Constants should be in uppercase with _ to separate the words and not camelCase.

Ryan-L98 commented on src/main/java/spark/parser/Parser.java:
```diff
@@ -0,0 +1,173 @@
+import spark.exceptions.SparkException;
+import spark.parser.commands.CommandKeyword;
+import spark.exceptions.formatexceptions.*;
+import spark.parser.commands.commandtypes.*;
```
Imported classes should always be listed explicitly 😢

Ryan-L98 commented on src/main/java/spark/parser/Parser.java:
```diff
@@ -0,0 +1,173 @@
+            return new ListCommand();
+
+        } else if (keyword == CommandKeyword.MARK) {
+            return new MarkCommand(getMarkParams(input, keyword));
```
I like how you gave each command a different parsing method, it really made the code look much cleaner! 💯

Ryan-L98 commented on src/main/java/spark/parser/commands/commandtypes/Command.java:
```diff
@@ -0,0 +1,15 @@
+public abstract class Command {
+    public abstract String execute(TaskList tasks, Ui ui, Storage storage);
+
+    public abstract boolean isExit();
```
Instead of having to override the `isExit` method in all the `Command` subclasses, you can consider using the java `instanceof` method since you do have an ExitCommand class to check if the command was to exit the application 😸

Ryan-L98 commented on src/main/java/spark/storage/Storage.java:
```diff
@@ -0,0 +1,94 @@
+    /** The location of the save-file on the user's hard-disk. */
+    private final Path filePath;
+    /** Represents the save-file on the user's hard-disk. */
+    private final File tasksFile;
```
Constant names must be all uppercase using underscore to separate words. 😫

Ryan-L98 commented on src/main/java/spark/tasks/tasktypes/Deadline.java:
```diff
@@ -0,0 +1,58 @@
+     *
+     * @param isDone
+     * @param title
+     * @param by
```
You can consider including some description about the params

Ryan-L98 commented on src/main/java/spark/tasks/tasktypes/Task.java:
```diff
@@ -0,0 +1,95 @@
+     *
+     * @throws TaskAlreadyUnMarked if the task is not complete.
+     */
+    public void unMark() throws TaskAlreadyUnMarked {
```
Nice unmark feature!

Ryan-L98 reviewed (commented):
Overall, looks good to merge! The code looks clean to me 😃 👍

zawedcvg commented on src/main/java/spark/parser/Parser.java:
```diff
@@ -0,0 +1,173 @@
+
+import spark.exceptions.SparkException;
+import spark.parser.commands.CommandKeyword;
+import spark.exceptions.formatexceptions.*;
```
Would be better if the import classes were imported separately

zawedcvg commented on src/main/java/spark/parser/commands/commandtypes/ExitCommand.java:
```diff
@@ -0,0 +1,24 @@
+    private String responseMessage;
+
+    @Override
+    public String execute(TaskList tasks, Ui ui, Storage storage) {
```
You could have this function as a part of the UI class. To improve abstraction.

zawedcvg commented on src/main/java/spark/parser/commands/commandtypes/FindTaskCommand.java:
```diff
@@ -0,0 +1,56 @@
+            }
+        }
+
+        responseMessage = results.toString();
```
You could carry out this function in the Ui class by passing results instead.

zawedcvg reviewed (commented):
Overall well written code following the coding style guidelines in most places.

jamesyeap committed:
DialogBox constructors: add assert statements

Constructor in DialogBox does not check that the input parameters are not null.

When null values are passed into DialogBox constructors, an unintelligible error message is printed to the console, making it difficult for developers to debug.

This situation may occur when the image to be included in the dialog box fails to be loaded by the program. This violates the basic assumption that all the required assets will be successfully loaded by the program.

To help make such an error easier to spot, let's add an explicit assertion to check that the image object is non-null.

jamesyeap committed:
Command response messages: encapsulate response messages inside a parent class CommandResponse

When a Command is executed by Spark, only a String object containing the response message is generated. This restricts metadata that can be specified about the response of the executed Command; such as whether it was produced as a result of a successful execution of the given Command, or if it was produced because of an exception produced when the Command was run.

Without this metadata, the user-interface cannot tailor the color of the chat box produced; ie it would not be possible to produce a red-colored chat box to indicate to the user that the input command could not be executed successfully.

To allow such additional information to be added, let's encapsulate the result of a Command into a class that not only contains the message to be shown to the user, and also other attributes of the result of the command's execution like whether it is an error, or if it is a warning that indicates some possibly undesirable impact resulting from its execution.

Doing so would provide additional visual feedback to the user about the result of the command's execution.

jamesyeap committed:
Merge pull request #1 from jamesyeap/branch-A-Assertions

Add assert statements to Spark

jamesyeap committed:
Merge branch 'master' into branch-A-CodeQuality

jamesyeap committed:
Merge pull request #2 from jamesyeap/branch-A-CodeQuality

Wrap result of command execution inside class CommandResponse

jamesyeap committed:
Edit Javadocs for DialogBox classes

jamesyeap committed:
Add feature to alert user if a duplicate Task is added

jamesyeap committed:
Add feature to display some message before exiting

jamesyeap committed:
Add example GitHub actions config file

jamesyeap committed:
Fix "gradlew checks" violation

jamesyeap committed:
Update .yml file

jamesyeap committed:
Remove Ui class

jamesyeap committed:
Add program title

jamesyeap committed:
Add UI screenshot to docs

jamesyeap committed:
Update user guide

jamesyeap committed:
Add a single section link to user guide

jamesyeap committed:
Change feature list in user guide

jamesyeap committed:
Update README.md

jamesyeap committed:
Add number of tasks remaining updates

jamesyeap committed:
Merge remote-tracking branch 'origin/master'
